Kernel32 Threading API

vatsan-madhavan opened the pull request:
- `CreateThread`, `CreateRemoteThread`, `CreateRemoteThreadEx`
- `CreateThreadFlags` enum
- Tests
- `FriendlyFlags.NativeInt` flag applicable to `IntPtr` and `UIntPtr` types that are used as a proxy for Win32 `SIZE_T` type. 
  - A future update will add code-gen for generating appropriate overloads leverage C# 9's `nint`/`nuint`. 
    - We need an issue for this. 
  - We probably also need an issue to inventory methods that have parameters or return-types that logically map to Wn32 `SIZE_T` and tag them with `FriendlyFlag.NativeInt`. 

Fixes #428

vatsan-madhavan committed:
Add `SafeThreadHandle`

vatsan-madhavan committed:
Add `CreateProcessFlags.STACK_SIZE_PARAM_IS_A_RESERVATION`

vatsan-madhavan committed:
Add `PInvoke.SIZE_T`

vatsan-madhavan committed:
Add `CreateThread`, `CreateRemoteThread`, and `CreateRemoteThreadEx`

vatsan-madhavan committed:
Add a test for `CreateThread`

vatsan-madhavan committed:
Add tests for `CreateRemoteThread` and `CreateRemoteThreadEx`

vatsan-madhavan committed:
Ensure SafeThreadHandle is release w/ 'using'

jkotas commented on src/Windows.Core/SIZE_T.cs:
```diff
@@ -0,0 +1,59 @@
+    /// the size of <see cref="IntPtr"/>
+    /// </remarks>
+    [StructLayout(LayoutKind.Sequential)]
+    public struct SIZE_T
```
Wrapper like this does not always behave the same way as direct `IntPtr` in native calling conventions. This wrapper won't always work as the `SIZE_T` type.

AArnott replied:
We've already added many APIs that take `SIZE_T` but just used `IntPtr` or `UIntPtr` for them so far. I think I like the idea of this struct if it makes it easier to deal with these values which are often intended to be consumable as an integer and `IntPtr` is really inconvenient for that. Is that the goal here?

jkotas replied:
The native integers feature in C# 9 is fixing the inconvenience of using `IntPtr` as an integer.

AArnott replied:
Awesome. Then do we need this struct at all?

vatsan-madhavan replied:
Don't think we need this struct - native ints (when they become available) should be great! @AArnott do you think you'd just generate overloads for `nint` for these kinds of methods with a new `[Friendly]` attribute ? Is there a helpful way to tag the parameter now so that it can be found easily in the future when scanning the codebase for appropriate `IntPtr`/`UIntPtr` parameters?

AArnott replied:
That's a good idea. Do you want to add a `NativeInt` flag to the `FriendlyFlags` enum now and we'll implement the code generator for it later?
But I wonder which target framework will benefit from `nint`. Just .NET Core 5.0, perhaps? Perhaps instead we should go ahead and define `SIZE_T` if it's easier to work with than `UIntPtr`, and we can program the code generator to generate `nint` overloads wherever it sees `SIZE_T` instead. That way, every single existing application today can actually benefit from some improvement both now and in the future.

vatsan-madhavan replied:
I believe C# 9 is only [supported on .NET 5](https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/#comment-6107). I will add `FriendlyFlags.NativeInt` - yes. 

As to `SIZE_T`, you're right - that would indeed be useful for `FriendlyFlags.NativeInt` code-generation for older TFM's. 

@jkotas, you wrote: 

>Wrapper like this does not always behave the same way as direct IntPtr in native calling conventions. This wrapper won't always work as the SIZE_T type.

What would be the failure mode here, and is there a way to write a workable `SIZE_T` struct definition that would mimic `UIntPtr` in the sense that it would have the same size as native `SIZE_T`/managed `UIntPtr` and can be treated like an integer in managed code?

jkotas replied:
> What would be the failure mode here

Hard to debug crash when `SIZE_T` is used in certain position. The specific situations where it would cause problems are platform specific. For example, believe `SIZE_T` return value cause problem on x86.

> is there a way to write a workable SIZE_T struct definition that would mimic UIntPtr 
 
There is not a way to write it with 100% fidelity.

jkotas replied:
> C# 9 is only supported on .NET 5

As the comment said - if you know what you are doing and can avoid the unsupported features, you should be able to use native integers even on .Net Standard 2.0 and .Net Framework. It is not artificially blocked.

AArnott replied:
> if you know what you are doing and can avoid the unsupported features, you should be able to use native integers even on .Net Standard 2.0 and .Net Framework. It is not artificially blocked.

That's how I use C# 8 on .NET Framework today. I'm fine with that. Do native ints use IntPtr under the covers, or a new type defined in .NET 5?

> Hard to debug crash when SIZE_T is used in certain position. The specific situations where it would cause problems are platform specific. For example, believe SIZE_T return value cause problem on x86.

That sounds like monsters I'd rather not invite to the party. I'm thinking maybe just using `UIntPtr` would be preferable.

jkotas replied:
> Do native ints use IntPtr under the covers

Yup

jkotas commented on src/Kernel32.Tests/storebanned/Kernel32Facts.cs:
```diff
@@ -944,6 +944,184 @@ public void SetHandleInformation_DoesNotThrow()
+                        hProcess.DangerousGetHandle(),
+                        &secAttrs,
+                        SIZE_T.Zero,
+                        new Kernel32.THREAD_START_ROUTINE(CreateThread_Test_ThreadMain),
```
There is nothing keeping this delegate alive while it is used by the unmanaged code that is the most likely reason for intermittent crashes.

You can read more about this at https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/callbackoncollecteddelegate-mda#symptoms

jkotas commented on src/Kernel32.Tests/storebanned/Kernel32Facts.cs:
```diff
@@ -944,6 +944,184 @@ public void SetHandleInformation_DoesNotThrow()
+        {
+            bInheritHandle = 1,
+            lpSecurityDescriptor = IntPtr.Zero,
+            nLength = Marshal.SizeOf<Kernel32.SECURITY_ATTRIBUTES>()
```
```suggestion
            nLength = sizeof(Kernel32.SECURITY_ATTRIBUTES)
```
I know that this is just a test code, but it is a few orders of magnide faster and more correct to use `sizeof` here.

jkotas replied:
(same in other places in this PR)

AArnott replied:
Interesting. Can you teach me more about when `sizeof` is OK to use for structs for interop purposes? When must `Marshal.SizeOf` be used? We use `Marshal.SizeOf` in many places in this repo so if `sizeof` is acceptable we can switch to it everywhere.

jkotas replied:
`Marshal.SizeOf` returns the size of the unmanaged view of the struct. `sizeof` returns the size of the managed view of the struct.

Unmanaged and managed views are identical for the blittable structs. It is better to use `sizeof` instead of `Marshal.SizeOf` for blittable structs. They return the same number and `sizeof` is much cheaper.

The code in this method is passing the managed view of the struct into the PInvoke without any marshaling. Unmanaged pointers like `SECURITY_ATTRIBUTES*` in this case are never marshalled. It is why I have said that it is more correct to use `sizeof` here.

AArnott replied:
Awesome. *all* of our structs in this repo are blittable, and our guidelines reinforce this. So it sounds like we should go through all of our use of `Marshal.SizeOf` and replace with `sizeof`. I'll open an issue to track.

AArnott commented on src/Kernel32/storebanned/Kernel32+SafeThreadHandle.cs:
```diff
@@ -0,0 +1,32 @@
+        /// <summary>
+        /// A <see cref="SafeHandle"/> for thread handles.
+        /// </summary>
+        public class SafeThreadHandle : SafeHandleZeroOrMinusOneIsInvalid
```
We already have `SafeObjectHandle` which is implemented in very nearly the same way, except that `-1` isn't considered invalid for `SafeObjectHandle`. Can we reuse that instead of defining a new SafeHandle type?

vatsan-madhavan replied:
That’s the 2nd (unstated) reason why I flipped the PR from active to draft :-). I realized this myself as I was working through the tests.

AArnott commented on src/Kernel32/storebanned/Kernel32.cs:
```diff
@@ -157,6 +157,25 @@ public static partial class Kernel32
+        /// Code that does this is common, but it can crash on 64-bit Windows.
+        /// </remarks>
+        [UnmanagedFunctionPointer(CallingConvention.Winapi)]
+        public delegate int THREAD_START_ROUTINE([In] IntPtr lpThreadParameter);
```
consider using `uint` here since it's `DWORD` in the header file. We sometimes use `int` instead but usually when it's likely that the value will need to be used in other .NET APIs that accept `int`. But in this case, it's an exit code which appears should always be unsigned.

AArnott commented on src/Kernel32/storebanned/Kernel32.cs:
```diff
@@ -157,6 +157,25 @@ public static partial class Kernel32
+        /// Code that does this is common, but it can crash on 64-bit Windows.
+        /// </remarks>
+        [UnmanagedFunctionPointer(CallingConvention.Winapi)]
+        public delegate int THREAD_START_ROUTINE([In] IntPtr lpThreadParameter);
```
We generally prefer using pointers instead of `IntPtr`. For p/invoke methods, we have code generators that automatically produce `IntPtr` overloads for us. In the case of delegates that's not possible, but I think we'd still want to keep with pointers since they're actually easier to work with in C# than `IntPtr`.
Also, we don't ever bother with `[In]` style attributes. If you know of a reason why that's important here, please enlighten me.

vatsan-madhavan replied:
I vaguely remember (now that you’ve jogged my memory) having discussed about [In] attribute before and concluding that I couldn’t find any purpose to it either. I’ll remove.

AArnott commented on src/Kernel32/storebanned/Kernel32.cs:
```diff
@@ -3026,5 +3045,166 @@ public static partial class Kernel32
+        /// </remarks>
+        [DllImport(api_ms_win_core_processthreads_l1_1_1, SetLastError = true)]
+        public static extern unsafe SafeThreadHandle CreateThread(
+            [Friendly(FriendlyFlags.In)] SECURITY_ATTRIBUTES* lpThreadAttributes,
```
Please add `| FriendlyFlags.Optional` to the first parameter since NULL is allowed per the documentation.

AArnott commented on src/Kernel32/storebanned/Kernel32.cs:
```diff
@@ -3026,5 +3045,166 @@ public static partial class Kernel32
+            THREAD_START_ROUTINE lpStartAddress,
+            IntPtr lpParameter,
+            CreateProcessFlags dwCreationFlags,
+            [Friendly(FriendlyFlags.Out)] int* lpThreadId);
```
Please add `| FriendlyFlags.Optional` to the last parameter since NULL is allowed per the documentation.

AArnott commented on src/Kernel32/storebanned/Kernel32.cs:
```diff
@@ -3026,5 +3045,166 @@ public static partial class Kernel32
+            SIZE_T dwStackSize,
+            THREAD_START_ROUTINE lpStartAddress,
+            IntPtr lpParameter,
+            CreateProcessFlags dwCreationFlags,
```
Is it really appropriate to reuse `CreateProcessFlags` for the `CreateThread` method?

vatsan-madhavan replied:
They share one enum value - CREATE_SUSPENDED. They also share the sentinel value `None`. 

That left me with a choice - expand CreateProcessFlags by 1 enum value (STACK_SIZE_PARAM_IS_A_RESERVATION), or create a new enum and add STACK_SIZE_PARAM_IS_A_RESERVATION, CREATE_SUSPENDED, and None to it. 

I don’t like the name `CreateProcessFlags` being used here, but these flags being in the same family (is the same enum) is reasonable from a Win32 standpoint (Process, threads, fibers go together). 

OTOH I don’t think thread creation fns would undergo fundamental breaking changes, and we could probably create `CreateThreadFlags` enum as a clean way if calling these methods. 

While pit-of-success design would normally compel me to instantiate a new enum type, calling CreateThread from managed code is an uncommon/advanced pattern and developers would likely know what they are doing. 

I could go either way. Would you rather I added a new enum?

AArnott replied:
Ya, I'd rather a new enum. Given `CreateProcessFlags` has so many flags that don't apply her, the overlap is small.

AArnott commented on src/Kernel32/storebanned/Kernel32.cs:
```diff
@@ -3026,5 +3045,166 @@ public static partial class Kernel32
+            [Friendly(FriendlyFlags.In)] SECURITY_ATTRIBUTES* lpThreadAttributes,
+            SIZE_T dwStackSize,
+            THREAD_START_ROUTINE lpStartAddress,
+            IntPtr lpParameter,
```
We tend to use `void*` where the native header uses it instead of `IntPtr`. you still get `IntPtr` overloads generated by the build but `void*` tends to be easier to use in C#.

AArnott commented on src/Kernel32/storebanned/Kernel32.cs:
```diff
@@ -3026,5 +3045,166 @@ public static partial class Kernel32
+        /// the process is created even if it refers to invalid or missing dynamic-link libraries (DLL).
+        /// </returns>
+        [DllImport(nameof(Kernel32), SetLastError = true)]
+        public static extern unsafe SafeThreadHandle CreateRemoteThread(
```
All/many of the same comments as the above method.

AArnott commented on src/Kernel32/storebanned/Kernel32.cs:
```diff
@@ -3026,5 +3045,166 @@ public static partial class Kernel32
+        /// If the function fails, the return value is NULL.To get extended error information, call <see cref="GetLastError"/>.
+        /// </returns>
+        [DllImport(nameof(Kernel32), SetLastError = true)]
+        public static extern unsafe SafeThreadHandle CreateRemoteThreadEx(
```
All/many of the same comments as the above method.

AArnott commented on src/Windows.Core/SIZE_T.cs:
```diff
@@ -0,0 +1,59 @@
+    /// This struct is intended for P/Invoke marshaling and must exactly match
+    /// the size of <see cref="IntPtr"/>
+    /// </remarks>
+    [StructLayout(LayoutKind.Sequential)]
```
nit: please drop this attribute. It's the C# default. Per (a relatively recent change in) our [contribution guidelines](https://github.com/dotnet/pinvoke/blob/master/CONTRIBUTING.md#structs).

AArnott commented on src/Windows.Core/SIZE_T.cs:
```diff
@@ -0,0 +1,59 @@
+        /// Gets special constant for (SIZE_T)0
+        /// </summary>
+        /// <remarks>
+        /// Intentionally not backed by a field to limit struct-size bloat; also
+        /// see remarks for <see cref="SIZE_T"/>.
```
How would a *static* field add bloat to the struct size?

AArnott commented on src/Windows.Core/SIZE_T.cs:
```diff
@@ -0,0 +1,59 @@
+        /// Casts to a ulong
+        /// </summary>
+        /// <param name="size">Value to be converted to ulong</param>
+        public static explicit operator ulong(SIZE_T size)
```
Is there some reason this is an *explicit* operator?
Can we add an implicit operator for converting `uint` and/or `ulong` to `SIZE_T` as well?

AArnott commented on src/Windows.Core/SIZE_T.cs:
```diff
@@ -0,0 +1,59 @@
+    [StructLayout(LayoutKind.Sequential)]
+    public struct SIZE_T
+    {
+        private IntPtr size;
```
Why not use `UIntPtr` instead, given `SIZE_T` is an unsigned integer? That would eliminate some casts you have to use below.

AArnott commented on src/Kernel32.Tests/storebanned/Kernel32Facts.cs:
```diff
@@ -944,6 +944,184 @@ public void SetHandleInformation_DoesNotThrow()
+        using var hProcess = Kernel32.GetCurrentProcess();
+
+        var result = false;
+        var gcHandle = GCHandle.Alloc(result);
```
Given `result` is a local variable, I wonder why you need to allocate a GC handle for it. Isn't its direct memory address sufficient? The runtime never rearranges memory on the stack, right? 
As written this method will not pop off the stack before the new thread exits, but even if it did, GCHandle can't guarantee this location on the stack will remain, right?

AArnott replied:
OK, looking at this more I guess it's implicitly boxing the `bool`, which is subtle, to say the least. Would it be more clear if you defined a `StrongBox<bool>` local variable and pinned that instead? Then you could access the `bool` any time instead of this odd creation of a local variable which you never use except to assign it later from the thing that *seemed* to pin it.

AArnott reviewed (changes requested):
Thanks for contributing!

vatsan-madhavan committed:
- Remove SafeThreadHandle - SafeObjectHandle has the same functionality.
- Remove SIZE_T - use UIntPtr instead; C# 9 can leverage nuint.
- Address review feedback
  - Replace IntPtr parameters with pointers.
  - Use FriendlyFlags.Optional where appropriate.
- Update tests.

vatsan-madhavan commented:
Addressed comments so far. Outstanding at this time: 

- Should we keep using `CreateProcessFlags` or  create a new `CreateThreadFlags` enum
- Add `FriendlyFlags.NativeInt` for future use. 
  - Bring back `SIZE_T` if possible to write one reliably.

vatsan-madhavan commented:
I couldn't  `void*` parameters with `FriendlyFlags.Optional` or `FriendlyFlags.Bidirectional`. I'm assuming that this is normal/expected?

AArnott commented:
> I couldn't tag void* parameters with FriendlyFlags.Optional or FriendlyFlags.Bidirectional. I'm assuming that this is normal/expected?

True, `void*` has no non-pointer alternative, so friendly attributes don't work on them.

AArnott commented:
>Should we keep using CreateProcessFlags or create a new CreateThreadFlags enum

Create a new one.

> Add FriendlyFlags.NativeInt for future use. Bring back SIZE_T if possible to write one reliably.

Future use option sounds good. The `SIZE_T` struct doesn't sound enticing given @jkotas' warning.

vatsan-madhavan committed:
Add FriendlyFlags.NativeInt

vatsan-madhavan committed:
Add CreateThreadFlags enum

vatsan-madhavan committed:
Keep a reference to the ThreadProc delegate in a static field.

AArnott commented on src/Kernel32.Tests/storebanned/Kernel32Facts.cs:
```diff
@@ -15,6 +15,7 @@
 
 public partial class Kernel32Facts
 {
+    private static unsafe Kernel32.THREAD_START_ROUTINE threadProc = new THREAD_START_ROUTINE(CreateThread_Test_ThreadMain);
```
Does being static prevent it from being moved by the GC?

Could you solve the problem with `GC.KeepAlive` or pinning of the delegate rather than using a static?

vatsan-madhavan replied:
Pinning seems unnecessary here. It seems that the interop-marshaler doesn't pass on the address of either the delegate or the function itself; rather what gets passed on is the address of the thunk that gets generated on the fly, and this address is in fact non-relocatable. I couldn't find any documentation re: the last point (non-relocatability property of the thunk) , but a scan of SO and dotnet/runtime sources strongly hints that this is the case. 

From [callbackOnCollectedDelegate MDA](https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/callbackoncollecteddelegate-mda): 
>When delegates are marshaled as function pointers, the runtime allocates a thunk that does the transition from unmanaged to managed. This thunk is what the unmanaged code actually calls before the managed delegate is finally invoked

For prevention from being moved by the GC (hindsight being 20/20, I should have looked here for answers to the AV I was encountering earlier) - 

>The number one mistake made when using delegates as unmanaged function pointers is to allow them to be garbage collected before unmanaged code is finished using them. Premature collection of delegates can result in several different symptoms, depending on the implementation of the unmanaged code using the function pointers. For example, it could cause an access violation, or it could cause your callback methods to never be called while the rest of the program runs normally. An easy way to prevent collection is to assign the delegate reference to a static field,

> Nathan, Adam. .NET and COM: The Complete Interoperability Guide (Kindle Locations 14677-14681). Pearson Education (US). Kindle Edition. 

I figured that a `static` is simpler since the delegate was being reused across multiple tests. `GC.KeepAlive` or `GCHandle.Alloc` might also work, I think. 

With this context, would you rather we didn't clutter `Kernel32Facts` with more top-level statics and rather move it over to local instances in each test-case?

AArnott replied:
What you have seems fine. I was mostly curious whether `GC.KeepAlive` would also work. My only concern for a test is that it can be used as a sample for how to accurately use an API, which yours does in one of a few likely workable fashions, so that's fine.

AArnott reviewed (approved):
Just one question about that test delegate

AArnott committed:
Merge branch 'master' into threadingapi

vatsan-madhavan commented:
Thanks for fixing up the branch and appreciate the feedback + help :thumbsup:

AArnott commented:
Thank you for the submission!
